File: 2707/solution.py
class Trie:
    def __init__(self):
        self.children = {}
        self.end = False


class Solution:
    def minExtraChar(self, s: str, dictionary: list[str]) -> int:
        root = Trie()
        for word in dictionary:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = Trie()
                node = node.children[char]
            node.end = True

        visited = {}

        def dp(index):
            if index == len(s):
                return 0
            if index in visited:
                return visited[index]
            res = 1 + dp(index + 1)

            node = root
            for i in range(index, len(s)):
                if s[i] not in node.children:
                    break
                node = node.children[s[i]]
                if node.end:
                    res = min(res, dp(i + 1))
                if res == 0:
                    break
            visited[index] = res
            return res

        return dp(0)

    # time O(d + s^2)
    # space O(d + s)

File: 2707/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    private Map<Integer, Integer> cache;
    private Trie root;

    public int minExtraChar(String s, String[] dictionary) {
        root = new Trie();
        cache = new HashMap<>();
        for (String word : dictionary) {
            root.insert(word);
        }
        return dp(0, s);
    }

    public int dp(int idx, String s) {
        if (cache.containsKey(idx))
            return cache.get(idx);
        if (idx == s.length()) {
            return 0;
        }

        int res = 1 + dp(idx + 1, s);
        Trie curr = root;
        for (int i = idx; i < s.length(); i++) {
            if (!curr.children.containsKey(s.charAt(i)))
                break;
            curr = curr.children.get(s.charAt(i));
            if (curr.word) {
                res = Math.min(res, dp(i + 1, s));
            }
        }
        cache.put(idx, res);
        return res;
    }
}

class Trie {
    boolean word;
    Map<Character, Trie> children;

    public Trie() {
        word = false;
        children = new HashMap<>();
    }

    public void insert(String word) {
        Trie root = this;
        for (int i = 0; i < word.length(); i++) {
            char curr = word.charAt(i);
            if (!root.children.containsKey(curr)) {
                root.children.put(curr, new Trie());
            }
            root = root.children.get(curr);
        }
        root.word = true;
    }
}

// time O(words * biggestWord + n*n)
// space O(words * biggestWord + n)